sdks/java/io/iceberg/src/main/java/org/apache/beam/io/iceberg/CombinedScanReader.java

```diff
@@ -0,0 +1,214 @@
+
+    files = new ArrayDeque<>();
+    files.addAll(task.files());
+
+    return advance();
+  }
+
+  @Override
+  public boolean advance() throws IOException {
+    Queue<FileScanTask> files =
+        checkStateNotNull(this.files, "files null in advance() - did you call start()?");
+    InputFilesDecryptor decryptor =
+        checkStateNotNull(this.decryptor, "decryptor null in adance() - did you call start()?");
+
+    // This is a lie, but the most expedient way to work with IcebergIO's
+    // which are not null-safe.
+    @SuppressWarnings("nullness")
+    org.apache.iceberg.@NonNull Schema project = this.project;
+
+    do {
+      // If our current iterator is working... do that.
+      if (baseIter != null && baseIter.hasNext()) {
+        current = baseIter.next();
+        return true;
+      }
+
+      // Close out the current iterator and try to open a new one
+      if (baseIter != null) {
+        baseIter.close();
+        baseIter = null;
+      }
+
+      LOG.info("Trying to open new file.");
+      @Nullable FileScanTask fileTask = null;
+      while (!files.isEmpty() && fileTask == null) {
+        fileTask = files.remove();
+        if (fileTask.isDataTask()) {
+          LOG.error("{} is a DataTask. Skipping.", fileTask.toString());
+          fileTask = null;
+        }
+      }
+
+      // We have a new file to start reading
+      if (fileTask != null) {
+        DataFile file = fileTask.file();
+        InputFile input = decryptor.getInputFile(fileTask);
+
+        @Nullable CloseableIterable<Record> iterable = null;
+        switch (file.format()) {
+          case ORC:
+            LOG.info("Preparing ORC input");
+            iterable =
+                ORC.read(input)
+                    .project(project)
+                    .createReaderFunc(
+                        fileSchema -> GenericOrcReader.buildReader(project, fileSchema))
+                    .filter(fileTask.residual())
+                    .build();
+            break;
+          case PARQUET:
+            LOG.info("Preparing Parquet input.");
+            iterable =
+                Parquet.read(input)
+                    .project(project)
+                    .createReaderFunc(
+                        fileSchema -> GenericParquetReaders.buildReader(project, fileSchema))
+                    .filter(fileTask.residual())
+                    .build();
+            break;
+          case AVRO:
+            LOG.info("Preparing Avro input.");
+            iterable =
+                Avro.read(input).project(project).createReaderFunc(DataReader::create).build();
+            break;
+          default:
+            throw new UnsupportedOperationException("Cannot read format: " + file.format());
+        }
+
+        if (iterable != null) {
+          baseIter = iterable.iterator();
+        }
+      } else {
+        LOG.info("We have exhausted all available files in this CombinedScanTask");
+      }
+
+    } while (baseIter != null);
+    return false;
```

Comment:
I think the protocol expect the `advance()` call to return true while there are elements. So we should set`current` and return `true` when a new iterator is set.

Reply:
I've rewritten it to be more clear. I think it can be refactored to be even simpler but for now it is working a bit. It returns true right after advancing a record (around line 108)